File: 2002_maximumProductOfTheLengthOfTwoPalindromicSubsequences/main.cpp
#include <string>

class Solution {
private:
    bool isPalindrome(std::string &s){
        int l = 0;
        int r = s.length() - 1;

        while (l < r) {
            if(s[l] != s[r])
                return false;
            l++;
            r--;
        }
        return true;
    }

    void dfs(int idx, std::string &s1, std::string &s2, std::string& s, int &res){
        if(idx >= s.length()) {
            if(isPalindrome(s1) && isPalindrome(s2)){
                int len = s1.length() * s2.length();
                res = std::max(res, len);
            }
            return;
        }
        
        char c = s[idx];

        // add the character in the first string
        s1.push_back(c);
        dfs(idx + 1, s1, s2, s, res);
        s1.pop_back();

        // add the character in the second string
        s2.push_back(c);
        dfs(idx + 1, s1, s2, s, res);
        s2.pop_back();

        // add character in no string
        dfs(idx + 1, s1, s2, s, res);
    }
public:
    int maxProduct(std::string s) {

        std::string s1 = "";
        std::string s2 = "";
        int idx = 0;
        int res = 0;
        dfs(idx, s1, s2, s, res);

        return res;
    }
};

File: 2002_maximumProductOfTheLengthOfTwoPalindromicSubsequences/main.py
class Solution:
    def maxProduct(self, s: str) -> int:
        N = len(s)
        palindrome = {}

        for mask in range(1, 1 << N):
            subseq = ""
            for i in range(N - 1, -1, -1):
                if mask & (1 << i):
                    subseq += s[i]
                if subseq == subseq[::-1]:
                    palindrome[mask] = len(subseq)
        res = 0
        for m1 in palindrome:
            for m2 in palindrome:
                if m1 & m2 == 0:
                    res = max(res, palindrome[m1] * palindrome[m2])

        return res